apps/src/blockly/addons/cdoSpritePointer.js

```diff
@@ -0,0 +1,112 @@
+// Helper functions for pointer block logic. In our pointer blocks (a.k.a.)
+// mini toolbox blocks) we "shadow" the image on the block from a source block.
+// This file contains functions to get and set the image on the pointer block.
+
+/**
+ * Get the image url for the given pointer block.
+ * @param {Block} block pointer block to get the image url for
+ * @param {*} pointerMetadataMap Object of the form {blockType: {imageSourceType: <string>, imageIndex: <number>}},
+ *  which maps pointer block types to the image source block type they can shadow and the index of the image to shadow.
+ * @param {string} imageSourceId Optional id of the image source block.
+ * @returns The url of the image that the pointer block should display, or an empty string if the block should display no image
+ */
+export function getPointerBlockImageUrl(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const pointerData = pointerMetadataMap[block.type];
+  if (!pointerData || !block.inputList || block.inputList.length === 0) {
+    return '';
+  }
+  let imageSource = undefined;
+  const mainWorkspace = Blockly.getMainWorkspace();
+  if (imageSourceId !== undefined) {
+    imageSource = mainWorkspace.getBlockById(imageSourceId);
+  }
+  if (!imageSource) {
+    const rootBlock = block.getRootBlock();
+    if (rootBlock.type === pointerData.imageSourceType) {
+      imageSource = rootBlock;
+    } else if (block.getRootBlock().type === block.type) {
+      const blockWorkspace = Blockly.Workspace.getById(block.workspace.id);
+      // If this block has itself as a root and is in a flyout workspace,
+      // it is a in a mini toolbox. A block can't be moved from one flyout to
+      // another, so if it's in a flyout, it is in its original flyout.
+      if (blockWorkspace && blockWorkspace.isFlyout && block.imageSourceId) {
+        imageSource = mainWorkspace.getBlockById(block.imageSourceId);
+      }
+    }
+  }
+  if (imageSource) {
+    return getImageUrlFromImageSource(imageSource, pointerData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
+  }
+}
+
+// Given a shadow block and an optional image source block id, update the image displayed
+// on the pointer block to match the image selected in the image source block. See getPointerBlockImageUrl
+// for details on how we find the image source block.
+export function updatePointerBlockImage(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const url = getPointerBlockImageUrl(block, pointerMetadataMap, imageSourceId);
+  changePointerImage(url, block);
+}
+
+// Get the image url from the image source block's inputList at the given index.
+// We find the image url by looking at the connection to the input at the given index.
+// If the image source does not have an image at the given index, return an empty string.
+function getImageUrlFromImageSource(imageSourceBlock, imageIndexOnSource) {
```

Comment:
I wonder if it would be worth providing an example value for `imageIndexOnSource`. With our current blocks, we'd always this expect to be 0 or 1, but it could theoretically also be higher... Is that right?

Reply:
Is the comment on the function not descriptive enough? ` Get the image url from the image source block's inputList at the given index.`. To me that implies it's an index in a list, which could be any number > 0.

Reply:
I'll update it to have a more full function comment (with `@param` notation) to make it clearer.

Reply:
Yeah, that’s true regarding an index in general. I was more wondering if there’s any reason we’d want to communicate that it will usually be 0 or 1.

Reply:
there's no reason it has to be 0 or 1, we could update the blocks to have many inputs if we wanted to, and shadow an input farther in the list.

Reply:
Yeah, that’s true. Feel free to disregard this feedback!